Add the ability to create a genesis block with the enrollment data.

MichaelKim20 opened the pull request:
Genesis block has frozen UTXOs of the validator's address.
It has the enrollment data using frozen UTXOs.

I extracted code from PR #845 written by @AndrejMitrovic .

Related to #822

AndrejMitrovic commented on source/agora/test/GenesisBlock.d:
```diff
@@ -16,83 +17,37 @@ module agora.test.GenesisBlock;
     scope(failure) network.printLogs();
     network.waitForDiscovery();
 
+    auto gen_tx = GenesisTransaction();
```
This is not the right genesis transaction. This is the main thread, and it doesn't have the custom genesis block that's set in `Base.d`. You should use `network.blocks`.

In fact all you have to do here is compare `node.getBlocksFrom(0, 1)[0]` with `network.blocks[0]`.

AndrejMitrovic replied:
So the `auto index = b.txs.enumerate.assumeSorted.find!(res => res[1] == gen_tx);` part should be replaced with just checking if the block is the same as the one in the APIManager (`network` variable).

MichaelKim20 replied:
That's a good suggestion.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -1028,3 +1034,70 @@ public bool containSameBlocks (APIS)(APIS nodes, size_t height)
+{
+    // 1 payment tx, the rest are freeze txs
+    Transaction[] txs;
+    txs ~= cast(Transaction)UnitTestGenesisTransaction;
```
Casting away immutable is bad practice. If we really need mutability we can do `UnitTestGenesisTransaction.serializeFull.deserializeFull!Transaction`.

MichaelKim20 replied:
That's a good suggestion. `serializeFull.deserializeFull`

AndrejMitrovic reviewed (changes requested):
Small changes requested, the rest looks good.

TrustHenry commented on source/agora/test/Base.d:
```diff
@@ -1028,3 +1034,77 @@ public bool containSameBlocks (APIS)(APIS nodes, size_t height)
+
+        Enrollment enroll;
+        const StartHeight = 1;
+        assert(enr.createEnrollment(utxo, StartHeight, enroll));
+        enrolls ~= enroll;
+    }
```
Genesis block begins with 2 validators.
So, Two enrollments have to be made.

AndrejMitrovic replied:
What? Where do you see this definition?

AndrejMitrovic replied:
The genesis block starts with whatever we define it to start with.

TrustHenry replied:
This is a test code. Consider this case:
https://github.com/bpfkorea/agora/issues/822#issue-617090208

AndrejMitrovic replied:
It's just an example. The code Michael wrote (which was taken from my PR) is generic and can create any number of enrollments.

> Consider this case:

This means "consider this example".

TrustHenry replied:
I think it would be nice to start with more than two enrollments.
Then we can test it more precisely.

MichaelKim20 replied:
Thank you for your good suggestion.

TrustHenry commented on source/agora/test/Base.d:
```diff
@@ -1028,3 +1034,77 @@ public bool containSameBlocks (APIS)(APIS nodes, size_t height)
+            merkle_root,
+            BitField!uint.init,
+            Signature.init,
+            enrolls.assumeUnique,
```
enrolls.sort!("a.utxo_key < b.utxo_key");
The enrollments should be aligned like this.

MichaelKim20 replied:
Thank you for your good suggestion.

TrustHenry reviewed (commented):
Overall, it's good.
In addition, unitTest should check the sorting of the enrolment.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -1028,3 +1034,79 @@ public bool containSameBlocks (APIS)(APIS nodes, size_t height)
+    txs ~= UnitTestGenesisTransaction.serializeFull.deserializeFull!Transaction;
+    Enrollment[] enrolls;
+
+    foreach (idx; 0 .. max(Enrollment.MinValidatorCount, key_pairs.length))
```
It would be much better to add an assertion that `key_pairs.length >= Enrollment.MinValidatorCount`.

If you generate random keypairs here, how will they be used by the Validators? Who will validate blocks [1..$]?

MichaelKim20 replied:
Only one validator is created when the value of TestConf.topology is NetworkTopology.OneValidator.
This is because the existing test code does not work.
So I created a dummy enrollment data. Is there any other good way to solve this?

AndrejMitrovic replied:
Just let it be one validator. There don't have to be two validators.

MichaelKim20 replied:
Thank you for the good suggestion.

AndrejMitrovic commented on source/agora/consensus/data/Enrollment.d:
```diff
@@ -51,6 +51,9 @@ public struct Enrollment
     public static immutable uint ValidatorCycle = 1008; // freezing period / 2
 
+    /// The minimum number of validators required to a create block
+    public static immutable uint MinValidatorCount = 2;
```
Why? This is introducing an arbitrary protocol change.

MichaelKim20 replied:
Yes, you're right. I'll remove this.

AndrejMitrovic commented:
I'll look at this PR more tomorrow.

But just to be more clear about this: If the test is wrong (it only has one validator), then the test should be fixed. We shouldn't add workaround code if our tests are broken. We should fix the tests instead. 

Because otherwise adding workarounds like this just means somebody else will have to fix the tests later on when we add block signature validation. 

So if we introduce a protocol rule that says "minimum 2 validators are enrolled at all times", it means we have to fix `NetworkTopology.OneValidator` and change it to  `NetworkTopology.TwoValidators`.

AndrejMitrovic commented on source/agora/test/GenesisBlock.d:
```diff
@@ -16,83 +17,36 @@ module agora.test.GenesisBlock;
 
-    // create spending txs refering to our own genesis block txs
-    auto txes = makeChainedTransactions(kp, null, 1, 40_000_000, genesis[1]);
+    assert(network.blocks[0].header.enrollments.length == 4);
+
+    auto index = network.blocks[0].txs
+        .enumerate.assumeSorted
+        .find!(res => res[1] == gen_tx);
+    assert(!index.empty);
```
We want to test that `NodeConfig.genesis_block` is read properly. 

```suggestion
    nodes.all!(node => node.getBlocksFrom(0, 1)[0] == network.blocks[0])
        .retryFor(2.seconds);
```

MichaelKim20 replied:
Yes, that's the more necessary test. 
I applied your suggestion.

MichaelKim20 committed:
Add the ability to create a genesis block with the enrollment data.

Genesis block has frozen UTXOs of the validator's address.
It has the enrollment data using frozen UTXOs.

Co-Authored-By: Andrej Mitrovic <andrej.mitrovich@gmail.com>
